File: src/libs/deck-of-cards/Deck.ts
import { Card, CardValues } from './Card';
import { Suits } from './Suit';

export class Deck {
  // TODO: Private!!
  public cards: Card[];

  constructor(shuffled = false) {
    this.cards = [];

    Suits.forEach(suit => {
      CardValues.forEach(value => {
        this.cards.push(new Card(suit, value));
      });
    });

    if (shuffled) {
      this.shuffle();
    }
  }

  public shuffle() {
    for (let i = this.cards.length - 1; i > 0; i--) {
      let j = Math.floor(Math.random() * (i + 1));

      let temp = this.cards[i];
      this.cards[i] = this.cards[j];
      this.cards[j] = temp;
    }
  }
}

File: flutter/lib/deck-of-cards/deck.dart
import 'card.dart';

class Deck {
  final List<PlayingCard> _deck;
  final List<PlayingCard> _drawn = [];

  Deck({bool shuffled = false})
      : _deck = List.generate(
          52,
          (index) => PlayingCard(
            suit: Suit.values[index % 4],
            value: (index % 13) + 1,
          ),
        ) {
    if (shuffled) {
      shuffle();
    }
  }

  void shuffle() {
    for (var i = _deck.length - 1; i > 0; i--) {
      var j = (DateTime.now().microsecondsSinceEpoch ~/ 1000) % (i + 1);

      var temp = _deck[i];
      _deck[i] = _deck[j];
      _deck[j] = temp;
    }
  }

  PlayingCard draw() {
    var drawnCard = _deck.removeAt(0);
    _drawn.add(drawnCard);

    return drawnCard;
  }

  bool isEmpty() {
    return _deck.isEmpty;
  }
}
